### PR TITLE
feat: Add support for http://example.com

### PR DESCRIPTION
**Pull request by Joey-1445601153**
fix:Add support for :// #15

### CODE REVIEW

File: endpoint_test.go [hunk 1]
```diff
@@ -81,12 +123,24 @@ func TestParse(t *testing.T) {
...
 			"tcp:127.0.0.1:xxx",
 			Endpoint{},
 			ErrInvalidValue,
 		},
+		{
+			"wrong port number in tcp with //",
+			"tcp://127.0.0.1:xxx",
+			Endpoint{},
+			ErrInvalidValue,
+		},
```
Comment (Xuanwo) [inline endpoint_test.go:197]: Please add test cases like:

- `file:///tmp/test`

File: endpoint_test.go [hunk 2]
```diff
@@ -33,6 +33,12 @@ func TestParse(t *testing.T) {
 			Endpoint{ProtocolHTTP, hostPort{"example.com", 80}},
 			nil,
 		},
+		{
+			"normal http with multi /",
+			"http://////example.com:80",
```
Comment (Xuanwo) [inline endpoint_test.go:38]: This should be invalid?

File: endpoint.go [hunk 3]
```diff
@@ -44,6 +44,10 @@ func Parse(cfg string) (p Endpoint, err error) {
 		// Handle file paths that contains ":" (often happens on windows platform)
 		//
 		// See issue: https://github.com/beyondstorage/go-endpoint/issues/3
+		s[1] = strings.TrimLeft(s[1], "/")
```
Comment (Xuanwo) [inline endpoint.go]: Compared to handle them in different protocol cases, how about handling all of them? Maybe we can operate on `s[1]` directly?

**Comment by 1445601153** [comment]
sorry, i delete a comment by accident.
for 
> @@ -33,6 +33,12 @@ func TestParse(t *testing.T) {
 			Endpoint{ProtocolHTTP, hostPort{"example.com", 80}},
 			nil,
 		},
+		{
+			"normal http with multi /",
+			"http://////example.com:80",
i tryed using address http://///74.125.128.160:80 to visit Google，it's valid.
Maybe it's valid for all website?

**Comment by Joey-1445601153** [comment]
@Xuanwo 
for
			"normal http with multi /",
			"http://////example.com:80",

I tried using address like http:////74.125.128.160:80 to visit google，it's valid.
I also create a simple local server. When I used address http:////127.0.0.1:8080/ to visit it, it was valid too.
So maybe it's valid for all?

**Comment by Xuanwo** [comment]
> http:////74.125.128.160:80

They may be valid for the browser, but they are not valid according to RFC3986: https://datatracker.ietf.org/doc/html/rfc3986#section-3

**Comment by Joey-1445601153** [comment]
@Xuanwo 
For the problem in endpoint.go, handling all of them together  may makes code chaos.
How about use a method like parseFile(imitate parseHostPort) to encapsulate these lines？

**Comment by Xuanwo** [comment]
> may makes code chaos

Please statement it more clear. Which problem could happen?

**Comment by Joey-1445601153** [comment]
> > may makes code chaos
> 
> Please statement it more clear. Which problem could happen?

If handling all of them together, we have to consider several situations in one method.The logic may be complex.One process may be entangled with another. It's hard to read and maintain.

**Comment by Xuanwo** [comment]
> we have to consider several situations in one method

Let's discuss them here now.

From my point of view, we only need to check like the following:

```go
if strings.HasPrefix(s[1], "//") {
    s[1] = s[1][2:]
}
```

Please correct me if there are other situations I missed.

**Comment by Joey-1445601153** [comment]
> > http:////74.125.128.160:80
> 
> They may be valid for the browser, but they are not valid according to RFC3986: https://datatracker.ietf.org/doc/html/rfc3986#section-3



> > we have to consider several situations in one method
> 
> Let's discuss them here now.
> 
> From my point of view, we only need to check like the following:
> 
> ```go
> if strings.HasPrefix(s[1], "//") {
>     s[1] = s[1][2:]
> }
> ```
> 
> Please correct me if there are other situations I missed.
If we just cut two headmost  '/' of s[1], it's OK.
when whole string is  http:///example.com, use s[1] = s[1][2:] will make the result of  ep.HTTP() is  http:///example.com again, i think the result is not  respecting the // notation.  According to RFC3986, it's invalid. It's valid for browser, wo also  have to consider about the different and how to process. To be honest, i 'm comfused about it.

<blockquote><div><strong><a href="https://datatracker.ietf.org/doc/html/rfc3986#section-3">rfc3986</a></strong></div><div>Uniform Resource Identifier (URI): Generic Syntax (RFC )</div></blockquote>

**Comment by Xuanwo** [comment]
> http:///example.com

We will return errors for invalid input. For example, we can return an error if `parseHostPort` got a string start with a `/`.

**Comment by Joey-1445601153** [comment]
> > http:///example.com
> 
> We will return errors for invalid input. For example, we can return an error if `parseHostPort` got a string start with a `/`.

Oh， so accutualy we can process s[1] before switch s[0].Then we don't have to change the method parseHostPort.
For http&https, we don‘t support input like http:///example.com.We just return an err for this kind of input.
And for file ,file:///tmp is valid, we encapsulate a method like parseFile to process the different between windows and linux.
Do i understand correctly?

**Comment by Xuanwo** [comment]
> Do i understand correctly?

100% correct.

Besides, we don't need a `parseFile` here, let the caller handle the difference.

**Comment by Xuanwo** [comment]
Much better! Thanks for your hard working!

**Pull request merged**